### ISSUE TITLE
BUG: Email Bounce/Delivered Status Data Cleanup

### ISSUE THREAD

**Issue by mjones-oddball** [department-of-veterans-affairs/notification-api#1645]
## Description
**Please note, this production cleanup cannot be done until after the deployment on 2/21. We need the fix to be deployed to stop this discrepancy from occurring. You can however, prep queries and test in the lowers before the deployment.**

Due to a bug in our status processing, we were falsely categorizing notifications as delivered. There is a use case where an email system can initially accept a notification (delivered), but later check the email and determine it to be undeliverable (soft or hard bounce). In both cases, if we receive a bounce after initially receiving delivered, we must update the status to to applicable failure status and reason.

See #1630 for more details on the bug fix.

- [x] Ticket is understood, and QA has been contacted (if the ticket has a QA label).

## Steps to Reproduce

1. Query our database and/or bigquery statistics data and search for notifications in a delivered status, containing a bounce related status reason.
2. Example Query for BQ:
```
SELECT
  date,
  service_name,
  template_name,
  status,
  status_reason
FROM
  `vsp-analytics-and-insights.platform_vanotify.prod-statistics`
WHERE
  status = "delivered"
  AND status_reason IS NOT null
```

## Impact/Urgency

High - skewing delivery results by about 0.2% overall

## Expected Behavior

**POC if you need any support is Evan who is familiar with cleanup processes**

Any email notification found in our database or bigquery with a status of delivered and a status_reason containing %bounce% should be updated to the appropriate failure status for hard or soft bounce.

Appropriate Statuses:
- Hard Bounce Status: `permanent-failure`
- Soft Bounce Status: `temporary-failure`

Checklist:

- [x] Create select count and update queries for the Prod Database and Big Query
    - [x] Test in Dev and dev-statistics. NOTE! There are other lower stat envs (perf and staging) that may have more data to test with. If you'd like to use those, that's fine. Not required to clean up all envs, but nice to have.
    - [x] Get approval from the Tech Lead

**DO NOT PROCEED UNTIL APPROVED.**

Production Database:
- [x] Perform the update in a [transaction block](https://www.postgresql.org/docs/current/tutorial-transactions.html) (do not commit)
- [x] Ensure the count of updated records matches the expected number of records to be updated, otherwise `ROLLBACK;` and raise this issue
- [x] If no issues with the counts, `COMMIT;` the query
- [x] Confirm with your select query that there are no remaining discrepancies

BigQuery prod-statistics table:
- [x] Perform the update in a transaction block
- [x] Ensure the count of updated records matches the expected number of records to be updated, otherwise `ROLLBACK;` and raise the issue
- [x] If no issues with the counts, `COMMIT;` the query
- [x] Confirm with your select query that there are no remaining discrepancies

### DB Queries
```
-- Hard Bounce
-- get the number of records you expect to update
SELECT count(*) FROM notifications as n
WHERE n.notification_type = 'email'
AND n.notification_status = 'delivered'
AND n.status_reason='Failed to deliver email due to hard bounce';

-- Soft Bounce
-- get the number of records you expect to update
SELECT count(*) FROM notifications as n
WHERE n.notification_type = 'email'
AND n.notification_status = 'delivered'
AND n.status_reason='Temporarily failed to deliver email due to soft bounce';

-- Hard Bounce
-- start the transaction
BEGIN;

-- run update and make sure updated count matches count above
UPDATE notifications as n
SET n.notification_status = 'permanent-failure'
WHERE n.notification_type = 'email'
AND n.notification_status = 'delivered'
AND n.status_reason='Failed to deliver email due to hard bounce';

-- if the numbers match commit the transaction,
-- otherwise rollback and see what went wrong
-- COMMIT; or ROLLBACK;

-- Soft Bounce
-- start the transaction
BEGIN;

-- run update and make sure updated count matches count above
UPDATE notifications as n
SET n.notification_status = 'temporary-failure'
WHERE n.notification_type = 'email'
AND n.notification_status = 'delivered'
AND n.status_reason='Temporarily failed to deliver email due to soft bounce';

-- if the numbers match commit the transaction,
-- otherwise rollback and see what went wrong
-- COMMIT; or ROLLBACK;
-------------------------------------------------------------

-- do the same for notification_history
-- may need to break this up into date ranges to reduce execution time
```

### BQ queries
```
-- Hard Bounce
-- get the number of records you expect to update
SELECT count(*) FROM `<env>-statistics`
WHERE channel_type = 'email'
AND status = 'delivered'
AND status_reason = 'Failed to deliver email due to hard bounce';

-- Soft Bounce
-- get the number of records you expect to update
SELECT count(*) FROM `<env>-statistics`
WHERE channel_type = 'email'
AND status = 'delivered'
AND status_reason = 'Temporarily failed to deliver email due to soft bounce';

-- Hard Bounce
-- start the transaction
BEGIN TRANSACTION;

-- run update and make sure updated count matches count above
UPDATE `<env>-statistics`
SET status = 'permanent-failure'
WHERE channel_type = 'email'
AND status = 'delivered'
AND status_reason = 'Failed to deliver email due to hard bounce';

-- if the numbers match commit the transaction,
-- otherwise rollback and see what went wrong
ROLLBACK TRANSACTION;
-- COMMIT TRANSACTION; -- when ready

-- Soft Bounce
-- start the transaction
BEGIN TRANSACTION;

-- run update and make sure updated count matches count above
UPDATE `<env>-statistics`
SET status = 'temporary-failure'
WHERE channel_type = 'email'
AND status = 'delivered'
AND status_reason = 'Temporarily failed to deliver email due to soft bounce';

-- if the numbers match commit the transaction,
-- otherwise rollback and see what went wrong
ROLLBACK TRANSACTION;
-- COMMIT TRANSACTION; -- when ready
```
## QA Considerations
_For QA to populate. Leave blank if QA is not applicable on this ticket._
- The next day, check that BigQuery prod-statistics results remains at 0 for the query under steps to reproduce

## Additional Info & Resources

- Another related ticket with query examples: https://github.com/department-of-veterans-affairs/vanotify-team/issues/1179
- [BigQuery](https://console.cloud.google.com/bigquery?project=vsp-analytics-and-insights&pli=1&ws=!1m4!1m3!3m2!1svsp-analytics-and-insights!2splatform_vanotify) stat tables:
<img width="288" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNnhVQXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--0b96fb03a20bc992d669f01f870b0a53282a544f/image.png" alt="image.png" />

**Comment by mjones-oddball** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/VA-Notify-620d21369d810a00146ed9c8/issues/department-of-veterans-affairs/notification-api/1645?planning-poker) with Zenhub @cris-oddball @EvanParish @k-macmillan @kalbfled @ldraney @mchlwellman

**Comment by mjones-oddball** [comment]
@k-macmillan I added a section for DB queries and BQ queries. Whoever takes this ticket can do a first review by testing them out in dev/perf before asking you to review and approve. However, if you want to take an initial look, feel free.

**Comment by EvanParish** [comment]
I updated the entries for all environments in Big Query. I also updated the Dev and Prod databases. I can do perf and staging on Monday if we want those to be updated as well. 

prod `notifications` table updates
<img width="806" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNW1JQXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--83d6fb6785c1b60162bfb575a62c5fbf5e4da787/image.png" alt="image.png" />

prod `notification_history` table hard bounce update
<img width="772" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNXFJQXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--768ce952a4adfc082cf539ea0ab73328423f819f/image.png" alt="image.png" />

prod `notification_history` table soft bounce update
<img width="871" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNXVJQXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--042ae8e71913424090b42a255fd86473d6b66793/image.png" alt="image.png" />

**Comment by EvanParish** [comment]
@cris-oddball The staging and perf environments may show some new hard / soft bounce counts when you check since I didn't update the database for those environments yet, but dev and prod should be clear of the error.
